File: Source/ARTOSReachability.m
#import <arpa/inet.h>
#import <ifaddrs.h>
#import <netdb.h>
#import <sys/socket.h>
#import <CoreFoundation/CoreFoundation.h>
#import <SystemConfiguration/SystemConfiguration.h>
#import <netinet/in.h>

#import "ARTOSReachability.h"
#import "ARTInternalLog.h"

typedef const void * __nonnull (* __nullable ARTNetworkReachabilityContextRetain)(const void * _Nullable info);

/// Global callback for network state changes
static void ARTOSReachability_Callback(SCNetworkReachabilityRef target, SCNetworkReachabilityFlags flags, void* info) {
    void (^callbackBlock)(SCNetworkReachabilityFlags) = (__bridge id)info;
    callbackBlock(flags);
}

@implementation ARTOSReachability {
    ARTInternalLog *_logger;
    NSString *_host;
    SCNetworkReachabilityRef _reachabilityRef;
    dispatch_queue_t _queue;
}

- (instancetype)initWithLogger:(ARTInternalLog *)logger queue:(dispatch_queue_t)queue {
    if (self = [super init]) {
        _logger = logger;
        _queue = queue;
    }
    return self;
}

- (void)listenForHost:(NSString *)host callback:(void (^)(BOOL))callback {
    [self off];
    _host = host;

    // This strategy is taken from Mike Ash's book "The Complete Friday Q&A: Volume III".
    // Article: https://www.mikeash.com/pyblog/friday-qa-2013-06-14-reachability.html
    
    __weak ARTOSReachability *weakSelf = self;
    void (^callbackBlock)(SCNetworkReachabilityFlags) = ^(SCNetworkReachabilityFlags flags) {
        ARTOSReachability *strongSelf = weakSelf;
        if (strongSelf) {
            BOOL reachable = (flags & kSCNetworkReachabilityFlagsReachable) != 0;
            ARTLogInfo(strongSelf->_logger, @"Reachability: host %@ is reachable: %@", strongSelf->_host, reachable ? @"true" : @"false");
            dispatch_async(strongSelf->_queue, ^{
                if (callback) {
                    callback(reachable);
                }
            });
        }
    };
    
    _reachabilityRef = SCNetworkReachabilityCreateWithName(NULL, [host UTF8String]);
    
    SCNetworkReachabilityContext context = {
        .version = 0,
        .info = (__bridge void *)(callbackBlock),
        .retain = (ARTNetworkReachabilityContextRetain)CFBridgingRetain,
        .release = CFRelease
    };
    if (SCNetworkReachabilitySetCallback(_reachabilityRef, ARTOSReachability_Callback, &context)) {
        if (SCNetworkReachabilityScheduleWithRunLoop(_reachabilityRef, CFRunLoopGetCurrent(), kCFRunLoopDefaultMode)) {
            ARTLogInfo(_logger, @"Reachability: started listening for host %@", _host);
        }
        else {
            ARTLogWarn(_logger, @"Reachability: failed starting listener for host %@", _host);
        }
    }
}

- (void)off {
    if (_reachabilityRef != NULL) {
        ARTLogInfo(_logger, @"Reachability: stopped listening for host %@", _host);
        SCNetworkReachabilityUnscheduleFromRunLoop(_reachabilityRef, CFRunLoopGetCurrent(), kCFRunLoopDefaultMode);
        SCNetworkReachabilitySetCallback(_reachabilityRef, NULL, NULL);
        CFRelease(_reachabilityRef);
        _reachabilityRef = NULL;
    }
    _host = nil;
}

- (void)dealloc {
    [self off];
}

@end

File: Sources/AblySwift/ARTOSReachability.swift
import Foundation
import SystemConfiguration

// Global typealias for network reachability context retain function  
typealias ARTNetworkReachabilityContextRetain = @convention(c) (UnsafeRawPointer?) -> UnsafeRawPointer?

// Helper class to wrap callback block for Core Foundation bridging
private class ARTReachabilityCallback {
    let callback: (SCNetworkReachabilityFlags) -> Void
    
    init(callback: @escaping (SCNetworkReachabilityFlags) -> Void) {
        self.callback = callback
    }
}

// swift-migration: original location ARTOSReachability.m, line 15
/// Global callback for network state changes
func ARTOSReachability_Callback(target: SCNetworkReachability, flags: SCNetworkReachabilityFlags, info: UnsafeMutableRawPointer?) {
    guard let info = info else { return }
    let callbackWrapper: ARTReachabilityCallback = Unmanaged.fromOpaque(info).takeUnretainedValue()
    callbackWrapper.callback(flags)
}

// swift-migration: original location ARTOSReachability+Private.h, line 3 and ARTOSReachability.m, line 20
internal class ARTOSReachability: NSObject, ARTReachability {
    private let logger: InternalLog
    private var host: String?
    private var reachabilityRef: SCNetworkReachability?
    private let queue: DispatchQueue
    
    // swift-migration: original location ARTOSReachability.m, line 27
    required init(logger: InternalLog, queue: DispatchQueue) {
        self.logger = logger
        self.queue = queue
        super.init()
    }
    
    // swift-migration: original location ARTOSReachability.m, line 35
    internal func listenForHost(_ host: String, callback: @escaping (Bool) -> Void) {
        off()
        self.host = host
        
        // This strategy is taken from Mike Ash's book "The Complete Friday Q&A: Volume III".
        // Article: https://www.mikeash.com/pyblog/friday-qa-2013-06-14-reachability.html
        
        weak var weakSelf = self
        let callbackBlock: (SCNetworkReachabilityFlags) -> Void = { flags in
            guard let strongSelf = weakSelf else { return }
            let reachable = (flags.rawValue & SCNetworkReachabilityFlags.reachable.rawValue) != 0
            ARTLogInfo(strongSelf.logger, "Reachability: host \(strongSelf.host ?? "") is reachable: \(reachable ? "true" : "false")")
            strongSelf.queue.async {
                callback(reachable)
            }
        }
        
        guard let reachability = SCNetworkReachabilityCreateWithName(nil, host) else {
            return
        }
        self.reachabilityRef = reachability
        
        // Create wrapper for the callback
        let callbackWrapper = ARTReachabilityCallback(callback: callbackBlock)
        let unmanagedCallback = Unmanaged.passRetained(callbackWrapper)
        
        var context = SCNetworkReachabilityContext(
            version: 0,
            info: unmanagedCallback.toOpaque(),
            retain: { info in
                return info
            },
            release: { info in
                let unmanaged = Unmanaged<ARTReachabilityCallback>.fromOpaque(info)
                unmanaged.release()
            },
            copyDescription: nil
        )
        
        if SCNetworkReachabilitySetCallback(reachability, ARTOSReachability_Callback, &context) {
            if SCNetworkReachabilityScheduleWithRunLoop(reachability, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue) {
                ARTLogInfo(logger, "Reachability: started listening for host \(host)")
            } else {
                ARTLogWarn(logger, "Reachability: failed starting listener for host \(host)")
            }
        }
    }
    
    // swift-migration: original location ARTOSReachability.m, line 74
    internal func off() {
        if let reachabilityRef = reachabilityRef {
            ARTLogInfo(logger, "Reachability: stopped listening for host \(host ?? "")")
            SCNetworkReachabilityUnscheduleFromRunLoop(reachabilityRef, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue)
            SCNetworkReachabilitySetCallback(reachabilityRef, nil, nil)
            self.reachabilityRef = nil
        }
        host = nil
    }
    
    // swift-migration: original location ARTOSReachability.m, line 85
    deinit {
        off()
    }
}
